### PR TITLE
cat/play: support directories

### PR DESCRIPTION
**Pull request by patapenka-alexey**
@TarantoolBot document
Title: Multiple directories with WAL files can be processed by `tt cat` and `tt play` commands

This patch adds support of multiple directories with WAL files to the `tt cat` and `tt play` commands.

```shell
$ tt cat /path/to/file.snap /path/to/file.xlog /path/to/dirs/
$ tt play <URI> /path/to/file.snap /path/to/file.xlog /path/to/dirs/
``` 

Closes https://github.com/tarantool/tt/issues/1020
Closes https://github.com/tarantool/tt/issues/1027

### CODE REVIEW

**Comment by oleg-jukovec** [comment]
Please, rebase on the master branch.

**Comment by oleg-jukovec** [comment]
Please, add a TarantoolBot message into a commit message too.

**Review by oleg-jukovec — APPROVED** [review]
LGTM!

File: cli/cmd/play.go [hunk 1]
```diff
@@ -75,8 +77,20 @@ func internalPlayModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
...
 	}
 
+	walFiles, err := util.CollectWALFiles(args[1:])
+	if err != nil {
+		return util.InternalError(
+			"Internal error: could not collect WAL files: %s",
+			version.GetVersion, err)
+	}
+
+	// Re-create args with the URI in the first index, and all founded files after
```
[the remark this answers is not in the document]
Comment (patapenka-alexey) [reply cli/cmd/play.go]: Fixed.

File: cli/cmd/play.go [hunk 2]
```diff
@@ -75,8 +77,20 @@ func internalPlayModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
 		return errors.New("it is required to specify an URI and at least one .xlog or .snap file")
```
Comment (themilchenko) [inline cli/cmd/play.go:77]: ... or dir?

File: cli/cmd/play.go [hunk 3]
```diff
@@ -75,8 +77,20 @@ func internalPlayModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
...
+	if err != nil {
+		return util.InternalError(
+			"Internal error: could not collect WAL files: %s",
+			version.GetVersion, err)
+	}
+
+	// Re-create args with the URI in the first index, and all founded files after.
+	var uriAndWalFiles []string
+	uriAndWalFiles = append(uriAndWalFiles, args[0])
+	uriAndWalFiles = append(uriAndWalFiles, walFiles...)
```
Comment (themilchenko) [inline cli/cmd/play.go:90]:
```suggestion
	uriAndWalFiles := append([]string{args[0]}, walFiles...)
```

File: cli/cmd/cat.go [hunk 4]
```diff
@@ -40,8 +40,10 @@ func NewCatCmd() *cobra.Command {
 				internalCatModule, args)
 			util.HandleCmdErr(cmd, err)
 		},
-		Example: "tt cat /path/to/xlog --timestamp 2024-11-13T14:02:36.818700000+00:00\n" +
-			"  tt cat /path/to/file.xlog /path/to/file.snap --timestamp=1731592956.818",
+		Example: "tt cat /path/to/file.snap /path/to/file.xlog /path/to/dirs/ " +
```
[the remark this answers is not in the document]
Comment (patapenka-alexey) [reply cli/cmd/cat.go:43]: It additionally notice user that multiple directories could be passed. We discussed about letter with @oleg-jukovec

File: cli/util/util.go [hunk 5]
```diff
@@ -967,3 +968,45 @@ func StringToTimestamp(input string) (string, error) {
 
 	return ts, nil
 }
+
+// CollectWALFiles globs files from args directories.
+func CollectWALFiles(paths []string) ([]string, error) {
+	if len(paths) < 1 {
+		return nil, errors.New("it is required to specify at least one .xlog or .snap file")
```
[the remark this answers is not in the document]
Comment (patapenka-alexey) [reply cli/util/util.go:975]: Ooops. Fixed.

File: cli/util/util.go [hunk 6]
```diff
@@ -967,3 +968,45 @@ func StringToTimestamp(input string) (string, error) {
...
+				return nil, err
+			} else {
+				for _, entry := range foundEntries {
+					collectedFiles = append(collectedFiles, filepath.Join(path, entry.Name()))
+				}
+				slices.SortFunc(collectedFiles, snapFilesFirst)
+			}
+		} else {
+			collectedFiles = append(collectedFiles, path)
+		}
```
[the remark this answers is not in the document]
Comment (patapenka-alexey) [reply cli/util/util.go:1008]: Looks cleaner. Agree.

[the remark this answers is not in the document]
Comment (themilchenko) [reply cli/util/util.go:1008]: I just found out how to make more cleaner

```suggestion
		if entry.IsDir() {
		        foundEntries, err := os.ReadDir(path)
		        if err != nil {
				return nil, err
			}

			for _, entry := range foundEntries {
				collectedFiles = append(collectedFiles, filepath.Join(path, entry.Name()))
			}
			slices.SortFunc(collectedFiles, snapFilesFirst)
			
			continue
		}
		
		collectedFiles = append(collectedFiles, path)
```

File: cli/cmd/cat.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/cmd/cat.go:43]: > It additionally notice user that multiple directories could be passed. We discussed about letter with @oleg-jukovec

I'm voting for `/path/to/dir/`.

File: cli/cmd/play.go (hunk 3 above)
Comment (patapenka-alexey) [reply cli/cmd/play.go:90]: Updated.

File: cli/util/util_test.go [hunk 7]
```diff
@@ -797,3 +797,58 @@ func TestStringToTimestamp(t *testing.T) {
...
+	snap1 := fmt.Sprintf("%s/%s", srcDir, "1.snap")
+	snap2 := fmt.Sprintf("%s/%s", srcDir, "2.snap")
+	xlog1 := fmt.Sprintf("%s/%s", srcDir, "1.xlog")
+	xlog2 := fmt.Sprintf("%s/%s", srcDir, "2.xlog")
+
+	tests := []struct {
+		name    string
+		input   []string
+		output  []string
+		wantErr bool
```
[the remark this answers is not in the document]
Comment (patapenka-alexey) [reply cli/util/util_test.go:816]: A few negative tests have been added.

File: cli/util/util.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (patapenka-alexey) [reply cli/util/util.go]: Updated.

File: cli/util/util_test.go [hunk 8]
```diff
@@ -797,3 +797,69 @@ func TestStringToTimestamp(t *testing.T) {
...
+			"",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var args []string
+			args = append(args, tt.input...)
+
+			result, err := CollectWALFiles(args)
```
[the remark this answers is not in the document]
Comment (patapenka-alexey) [reply cli/util/util_test.go]: Ooops. Fixed.

File: cli/util/util_test.go [hunk 9]
```diff
@@ -797,3 +797,69 @@ func TestStringToTimestamp(t *testing.T) {
...
+		},
+		{
+			"mix",
+			[]string{srcDir, "util_test.go"},
+			[]string{snap1, snap2, xlog1, xlog2, "util_test.go"},
+			"",
+		},
+	}
+
+	for _, tt := range tests {
```
[the remark this answers is not in the document]
Comment (patapenka-alexey) [reply cli/util/util_test.go]: Just copied-pasted from existing code. I changed it to `test`.

**Review by themilchenko — APPROVED** [review]
Looks nice, thank you!

**Comment by patapenka-alexey** [comment]
I removed `dirs` from commit message.

**Pull request merged**